olp-cpp-sdk-core/src/network2/curl/NetworkProtocolCurl.cpp

```diff
@@ -0,0 +1,1230 @@
+  }
+
+  // handles setup
+  std::shared_ptr<NetworkProtocolCurl> that = shared_from_this();
+  for (int i = 0; i < kTotalHandleCount; i++) {
+    if (i < kStaticHandleCount) {
+      handles_[i].handle = curl_easy_init();
+    } else {
+      handles_[i].handle = nullptr;
+    }
+    handles_[i].index = i;
+    handles_[i].in_use = false;
+    handles_[i].self = that;
+  }
+
+  // start worker thread
+  thread_ = std::thread(&NetworkProtocolCurl::Run, this);
+
+  std::unique_lock<std::mutex> lock(event_mutex_);
+  event_condition_.wait(lock,
+                        [this] { return state_ == WorkerState::Started; });
+
+  return true;
+}
+
+void NetworkProtocolCurl::Deinitialize() {
+  // Stop worker thread
+  if (state_ != WorkerState::Started) {
+    EDGE_SDK_LOG_DEBUG(kLogTag, "Already deinitialized");
+    return;
+  }
+  {
+    std::lock_guard<std::mutex> lock(event_mutex_);
+    state_ = WorkerState::Stopping;
+    event_condition_.notify_one();
+  }
+
+  std::lock_guard<std::mutex> initLock(init_mutex_);
+  if (thread_.get_id() != std::this_thread::get_id()) {
+    event_condition_.notify_all();
+    thread_.join();
+  } else {
+    // We are trying to stop the very thread we are in. This is not recommended,
+    // but we try to handle it gracefully. This could happen by calling from one
+    // of the static functions (rxFunction or headerFunction) that was passed to
+    // the cURL as callbacks.
+    thread_.detach();
+  }
+}
+
+void NetworkProtocolCurl::Teardown() {
+#if (defined NETWORK_HAS_PIPE) || (defined NETWORK_HAS_PIPE2)
+  char tmp = 1;
+  if (write(pipe_[1], &tmp, 1) < 0) {
+    EDGE_SDK_LOG_INFO(kLogTag, "deinitialize - failed to write pipe " << errno);
+  }
+#endif
+
+  // handles teardown
+  std::vector<std::pair<RequestId, Network::Callback> >
+      completedMessages;
+  {
+    std::lock_guard<std::mutex> lock(event_mutex_);
+    events_.clear();
+
+    for (int i = 0; i < kTotalHandleCount; i++) {
+      if (handles_[i].handle) {
+        if (handles_[i].in_use) {
+          curl_multi_remove_handle(curl_, handles_[i].handle);
+          completedMessages.emplace_back(handles_[i].id, handles_[i].callback);
+        }
+        curl_easy_cleanup(handles_[i].handle);
+        handles_[i].handle = nullptr;
+        handles_[i].self.reset();
+      }
+    }
+  }
+
+  // cURL teardown
+  curl_multi_cleanup(curl_);
+  curl_ = nullptr;
+
+#ifdef NETWORK_HAS_OPENSSL
+  // OpenSSL teardown
+  CRYPTO_set_id_callback(nullptr);
+  CRYPTO_set_locking_callback(nullptr);
+  gSslMutexes = nullptr;
+  ssl_mutexes_.reset();
+#endif
+
+#if (defined NETWORK_HAS_PIPE) || (defined NETWORK_HAS_PIPE2)
+  close(pipe_[0]);
+  close(pipe_[1]);
+#endif
+
+  // Handle completed messages
+  if (!completedMessages.empty()) {
+    for (auto& pair : completedMessages) {
+      pair.second(network2::NetworkResponse()
+        .WithRequestId(pair.first)
+        .WithStatus(static_cast<int>(ErrorCode::OFFLINE_ERROR))
+        .WithError("Offline"));
```

Comment:
```suggestion
        .WithError("Offline: network is deintialized"));
```
I guess it is better to extend the error string that the error happened during teardown (and to distinguish this case from regular Offline).